site/source/docs/porting/pthreads.rst

```diff
@@ -43,17 +43,82 @@ but is unrelated. That flag does not use pthreads or SharedArrayBuffer, and
 instead uses a plain Web Worker to run your main program (and postMessage to
 proxy messages back and forth).
 
+Proxying
+========
+
+The Web allows certain operations to only happen from the main browser thread,
+like interacting with the DOM. As a result, various operations are proxied to
+the main browser thread if they are called on a background thread. See
+`bug 3495 <https://github.com/emscripten-core/emscripten/issues/3495>`_ for
+more information and how to try to work around this until this. To check which
+operations are proxied, you can look for the function's implementation in
+the JS library (``src/library_*``) and see if it is annotated with
+``__proxy: 'sync'`` or ``__proxy: 'async'``; however, note that the browser
+itself proxies certain things (like some GL operations), so there is no
+general way to be safe here (aside from not blocking on the main browser
+thread).
+
+In addition, Emscripten currently has a simple model of file I/O only happening
+on the main application thread (as we support JS plugin filesystems, which
+cannot share memory), which is another set of operations that are proxied.
+
+Normally you don't need to worry about this proxying, as it is fully automatic,
```

Comment:
"Normally you don't need to worry about this proxying" I disagree with this statement. Every web developer (especially those porting things from other platforms) has to worry about the main thread, and every developer who uses threads has to worry about which thread things happen on. Basically 100% of the users we've worked with who use threads run into some issue or other (which is why PROXY_TO_PTHREAD exists).  That's just reality and I don't think it's going to change. Also the "however, " clause proves that this statement isn't true anyway. That's really the reason I've been pushing to make proxying more clear and understandable in the first place.

Reply:
Fair enough. Changing to `Proxying can cause problems in certain cases, see the section on blocking right after this one.`

Reply:
could just be "section on blocking below". Although, isn't proxying the solution to the blocking problems, rather than the cause?

Reply:
https://d.justpo.st/media/images/2013/10/dc86338697d8e4e4364da35a762535ed.jpg

It is kind of both... e.g. we proxy stdio to the main thread which is helpful in some ways, but can cause new deadlocks.